src/network/matrix/matrix.ts

```diff
@@ -0,0 +1,175 @@
+        content: object
+        origin_server_ts: number
+        unsigned: unknown & {
+            age: number
+        }
+        event_id: string
+        room_id: string
+    }
+}
+export class MatrixMessage {
+    public userID: null | string = null
+    constructor(public matrixClient: MatrixClient) {
+        matrixClient
+            .startClient({})
+            .then(() => matrixClient.getUserId())
+            .then(id => (this.userID = id))
+
+        matrixClient.on(
+            'Room.timeline',
+            <T extends keyof MatrixMessageTypes>(
+                event: PatchedEvent,
+                _room: Room,
+                _eventTimelineSet: EventTimelineSet,
+                _room2: Room,
+            ) => {
+                const e = event.event
+                const data: MatrixMessageEvent<T> = {
+                    type: e.type as T,
+                    meta: {
+                        matrix: {
+                            room_id: e.room_id,
+                            room_name: this.reverseCache.get(e.room_id),
+                            event_id: e.event_id,
+                            origin_server_ts: e.origin_server_ts,
+                            sender: e.sender,
+                        },
+                    },
+                    // TODO: validate it with JSON schema
+                    payload: e.content as MatrixMessageTypes[T],
+                }
+                this.mitt.emit(e.type, data)
+            },
+        )
+    }
+    async emit<T extends keyof MatrixMessageTypes>(
+        room: { type: 'alias'; alias: string } | { type: 'id'; id: string },
+        eventName: T,
+        data: MatrixMessageTypes[T],
+    ) {
+        let id = ''
+        if (room.type === 'alias') id = await this.lookupRoomAlias(room.alias)
+        else if (room.type === 'id') id = room.id
+        await this.matrixClient.sendEvent(id, eventName, data, undefined!, undefined!)
+    }
+    on<T extends keyof MatrixMessageTypes>(
+        event: keyof MatrixMessageTypes,
+        f: MatrixMessageListener<T>,
+        options?: { ignoreMyself?: boolean },
+    ) {
+        const _f: typeof f = event => {
+            if (event.meta.matrix.sender === this.userID && options?.ignoreMyself) return
+            f(event)
+        }
+        this.mitt.on(event, _f)
+        return () => this.mitt.off(event, _f)
+    }
+    private mitt = mitt()
+    private cache = new Map<string, string>()
+    private reverseCache = new Map<string, string>()
+    private async lookupRoomAlias(alias: string): Promise<string> {
+        if (this.cache.has(alias)) return this.cache.get(alias)!
+        // @ts-ignore
+        const x: { room_id: string; servers: string[] } = await this.matrixClient.getRoomIdForAlias(alias)
+        this.cache.set(alias, x.room_id)
+        this.reverseCache.set(x.room_id, alias)
+        return x.room_id
+    }
+}
+export async function createMatrixMessage(username: string, password: string) {
```

Comment:
Refactor these startup code as unit testing cases?

Reply:
I didn't catch you r meaning

Reply:
It looks just some test code here.

Reply:
Never mind. Too many things need to mock here.